instrumentation/kafka-clients/src/main/java/brave/kafka/clients/TracingProducer.java

```diff
@@ -0,0 +1,114 @@
+package brave.kafka.clients;
+
+import brave.Span;
+import brave.Span.Kind;
+import brave.Tracing;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+class TracingProducer<K, V> implements Producer<K, V> {
+
+  private static final KafkaPropagation.ProducerInjector injector =
+      new KafkaPropagation.ProducerInjector();
+
+  private final Tracing tracing;
+  private final Producer<K, V> wrappedProducer;
+
+  TracingProducer(Tracing tracing, Producer<K, V> producer) {
+    this.wrappedProducer = producer;
+    this.tracing = tracing;
+  }
+
+  @Override
+  public void initTransactions() {
+    wrappedProducer.initTransactions();
+  }
+
+  @Override
+  public void beginTransaction() {
+    wrappedProducer.beginTransaction();
+  }
+
+  @Override
+  public void commitTransaction() {
+    wrappedProducer.commitTransaction();
+  }
+
+  @Override
+  public void abortTransaction() {
+    wrappedProducer.abortTransaction();
+  }
+
+  /**
+   * Send with a callback is always called for KafkaProducer. We do the same here to enable
+   * tracing.
+   */
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+    return this.send(record, null);
+  }
+
+  /**
+   * We wrap the send method to add tracing.
+   */
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+    Span span = tracing.tracer().nextSpan();
+
+    if (record.key() != null) {
+      span.tag(KafkaTags.KAFKA_KEY_TAG, record.key().toString());
+    }
+    span.tag(KafkaTags.KAFKA_TOPIC_TAG, record.topic());
+
+    tracing.propagation()
+        .injector(injector)
+        .inject(span.context(), record);
+
+    Future<RecordMetadata> recordMetadataFuture = wrappedProducer.send(record, callback);
```

Comment:
wrap the callback and in your wrapped one, call span.finish() or record an error tag, then call span.finish() accordingly